llvm/lib/Target/AIE/AIEHazardRecognizer.cpp

```diff
@@ -399,31 +412,47 @@ auto toHazardType(bool Conflict) {
 // recognizing alternatives
 ScheduleHazardRecognizer::HazardType
 AIEHazardRecognizer::getHazardType(const MCInstrDesc &Desc,
+                                   MemoryBankBits MemoryBanks,
                                    const int DeltaCycles) {
-  return toHazardType(checkConflict(
-      Scoreboard, ItinData, Desc.getSchedClass(),
+  return getHazardType(
+      Desc.getSchedClass(),
       getSlotSet(Desc, *TII->getFormatInterface(), IgnoreUnknownSlotSets),
-      DeltaCycles, FUDepthLimit));
+      MemoryBanks, DeltaCycles);
 }
 
 ScheduleHazardRecognizer::HazardType
 AIEHazardRecognizer::getHazardType(unsigned SchedClass, SlotBits SlotSet,
+                                   MemoryBankBits MemoryBanks,
                                    int DeltaCycles) {
-  return toHazardType(checkConflict(Scoreboard, ItinData, SchedClass, SlotSet,
-                                    DeltaCycles, FUDepthLimit));
+  return toHazardType(checkConflict(
+      Scoreboard, ItinData, SchedClass, SlotSet, MemoryBanks,
+      TII->getMemoryCycles(SchedClass), DeltaCycles, FUDepthLimit));
 }
 
 bool AIEHazardRecognizer::checkConflict(
     const ResourceScoreboard<FuncUnitWrapper> &Scoreboard,
     const InstrItineraryData *ItinData, unsigned SchedClass, SlotBits SlotSet,
-    int DeltaCycles, std::optional<int> FUDepthLimit) {
+    MemoryBankBits MemoryBanks,
+    std::optional<std::vector<int>> MemoryAccessCycles, int DeltaCycles,
+    std::optional<int> FUDepthLimit) {
   assert(Scoreboard.isValidDelta(DeltaCycles));
 
   // Verify format hazards
   FuncUnitWrapper EmissionCycle(/*Req=*/0, /*Res=*/0, SlotSet);
   if (EmissionCycle.conflict(Scoreboard[DeltaCycles]))
     return true;
 
+  // Verify memory bank hazards
+  if (MemoryAccessCycles.has_value()) {
```

Comment:
See comment before, I think we should pass `const SmallVector<int, 2> &MemoryAccessCycles` here, and just iterate over the elements.